cpuset/cpuset.go

```diff
@@ -0,0 +1,318 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cpuset
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+// Builder is a mutable builder for CPUSet. Functions that mutate instances
+// of this type are not thread-safe.
+type Builder struct {
+	result CPUSet
+	done   bool
+}
+
+// NewBuilder returns a mutable CPUSet builder.
+func NewBuilder() *Builder {
+	return &Builder{
+		result: CPUSet{
+			cpuIDs: map[int]struct{}{},
+		},
+	}
+}
+
+// Add adds the supplied CPU IDs to the result. Calling Add after calling
+// Result has no effect. CPU IDs are sometimes known as processor IDs.
+// They are distinct from core IDs, which identify a physical core, because
+// some systems have multiple vCPUs on each core. These vCPUs would each
+// have their own CPU ID.
+func (b *Builder) Add(cpuIDs ...int) {
+	if b.done {
+		return
+	}
+	for _, cpuID := range cpuIDs {
+		b.result.cpuIDs[cpuID] = struct{}{}
+	}
+}
+
+// Result returns the result CPUSet containing all CPU IDs that were
+// previously added to this builder. Subsequent calls to Add have no effect.
+func (b *Builder) Result() CPUSet {
+	b.done = true
+	return b.result
+}
+
+// CPUSet is a thread-safe, immutable set-like data structure for CPU IDs.
```

Comment:
that seems reasonable to me. I have to admit, at this point, I'm not super thrilled about waiting an unbounded amount of time for a dependency to land though.

Reply:
I found that the 'sets' import issue is unclaimed: https://github.com/kubernetes/utils/issues/9, so I can probably just execute that import too, if you want.

Reply:
There's a generic sets PR that we want instead, I think, but that can't go in for a while because of older supported releases.  What I think I would do here is make this API use all the same names and signatures as the set API, and if/when it lands we can use it isdie here.  We still want a type for String(), if nothing else, right?

But I should be clear - I am asking, not stating, about immutable.  Is it an important property?  If so, OK.  If not, the whole Builder pattern can go away.

Reply:
As the maintainer of the `cpumanager`, I can definitively say that immutability is not important in this abstraction. I would much prefer it to use the standard set implementation with extensions only for the `Parse()` and `String()` methods.

Reply:
@thockin Are you referring to the `sets` implementation in `apimachinery/pkg/util/sets` or elsewhere? The PR for the generic sets implementation has already landed there.

Is it reasonable to import from there into this new `kubernetes/utils/cpuset` package to create a `CPUSet` type which wraps / extends `sets.New[int]`.

Or did you have something else in mind?